compiler/hash-source/src/location.rs

```diff
@@ -105,3 +108,117 @@ impl SourceLocation {
         Self { id: self.id, span: self.span.join(other.span) }
     }
 }
+
+/// Represents a position within a source using a `row` and `column`  
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub struct RowCol {
+    /// The row number, indexing starts from `0`, but when printed, one is
+    /// always added as most editors display rows beginning from `1`.
+    pub row: usize,
+    /// The column number, indexing starts from `0`, but when printed, one is
+    /// always added as most editors display rows beginning from `1`.
+    pub column: usize,
+}
+
+impl Display for RowCol {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}:{}", self.row + 1, self.column + 1)
+    }
+}
+
+/// [RowColSpan] is a data structure that is equivalent to [Span] but uses rows
+/// and columns to denote offsets within the source file. [RowColSpan] is only
+/// re-used when specific line numbers need to be reported, this shouldn't be
+/// used for general purpose storage of positions of source items.
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub struct RowColSpan {
+    /// The starting position of the [RowColSpan].
+    pub start: RowCol,
+    /// The end position of the [RowColSpan].
+    pub end: RowCol,
+}
+
+impl RowColSpan {
+    /// Create a new [RowColSpan] from a `start` and `end`.
+    pub fn new(start: RowCol, end: RowCol) -> Self {
+        Self { start, end }
+    }
+
+    /// Get the associated rows with the start and end of the [RowColSpan].
+    pub fn rows(&self) -> (usize, usize) {
+        (self.start.row, self.end.row)
+    }
+
+    pub fn columns(&self) -> (usize, usize) {
+        (self.start.column, self.end.column)
+    }
+}
+
+impl Display for RowColSpan {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if self.start == self.end {
+            write!(f, "{}", self.start)
+        } else {
+            write!(f, "{}-{}", self.start, self.end)
+        }
+    }
+}
+
+/// Function to compute a row and column number from a given source string
+/// and an offset within the source. This will take into account the number
+/// of encountered newlines and characters per line in order to compute
+/// precise row and column numbers of the span.
+pub fn compute_row_col_from_offset(offset: usize, source: &str, non_inclusive: bool) -> RowCol {
+    let source_lines = source.split('\n');
+
+    let mut bytes_skipped = 0;
+    let mut total_lines: usize = 0;
+    let mut last_line_len = 0;
+
+    let mut line_index = None;
+    for (line_idx, line) in source_lines.enumerate() {
+        // One byte for the newline
```

Comment:
what about windows ?

Reply:
I think the best approach in the future is to `normalise` sources when they are first to read in so that we don't have to worry about them at all after reading them in.